services/user/Tokens/plugin/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+#[allow(warnings)]
+mod bindings;
+
+use bindings::common::plugin::{server, types as CommonTypes};
+use bindings::exports::component::tokens::types as Wit;
+use bindings::exports::component::tokens::{intf::Guest as Intf, transfer::Guest as Transfer};
+use psibase::services::tokens as Wrapper;
+use psibase::AccountNumber;
+use query::token_detail::fetch_token;
+
+mod errors;
+use errors::ErrorType;
+struct Component;
+
+use psibase::fracpack::Pack;
+
+mod query {
+    pub mod token_detail;
+}
+
+enum TokenType {
+    Number(u32),
+    Symbol(String),
+}
+
+fn identify_token_type(token_id: String) -> Result<TokenType, CommonTypes::Error> {
+    use TokenType::{Number, Symbol};
+
+    let first_char = token_id
+        .chars()
+        .next()
+        .ok_or(ErrorType::InvalidTokenId.err("token code is empty"))?;
+
+    Ok(if first_char.is_ascii_digit() {
+        Number(
+            token_id
+                .parse::<u32>()
+                .map_err(|_| ErrorType::InvalidTokenId.err("failed to parse token_id to u32"))?,
+        )
+    } else {
+        Symbol(token_id)
+    })
+}
+
+fn token_code_to_id(token_id: Wit::TokenId) -> Result<u32, CommonTypes::Error> {
+    let parsed = identify_token_type(token_id)?;
+    match parsed {
+        TokenType::Number(number) => Ok(number),
+        TokenType::Symbol(_) => {
+            unimplemented!("Dunno how to lookup a token by symbol")
```

Comment:
> Better to return an error that can be handled rather than panic

Feels funky to me, we typically write error handling for errors that **might** happen, do we want to write error handling for things we know **won't** happen? 

I could argue this has a case for stuff deployed live where UI devs are reaching out to plugin functions with failures in v1 but success in v1.1, but feels a bit over the top for now.

Reply:
Added